File: CopilotAI/DSLK.cpp
#include<stdio.h>
#include<iostream>
struct node
{
    int data;
    node *next;
};
//typedef struct node node;
node* makeNode(int x){
    node *newNode = new node();
    newNode->data = x;
    newNode->next = NULL;
    return newNode;
}
void duyet(node *head){
    while(head != NULL){
       std:: cout << head->data << " ";
        head = head->next;      //  head->next dia chi cua node ke tiep
    }
}
// In ra so luong node trong DSLK
int count(node *head){
    int count = 0;
    while(head != NULL){
        count++;
        head = head->next;
    }
    return count;
}
// Them 1 node vao dau  DSLK
void addHead(node **head, int x){
    node *newNode = makeNode(x);
    newNode->next = *head;
    *head = newNode;
}
// Add 1 node vao cuoi DSLK
void addTail(node **head, int x){
    node *newNode = makeNode(x);
    if(*head == NULL){
        *head = newNode;
    }
    else{
        node *temp = *head;
        while(temp->next != NULL){
            temp = temp->next;
        }
        temp->next = newNode;
    }
}   

int main(){
    node *head = NULL;
    for (int i = 1; i <= 10; i++ )
{
   // addTail(&head, i);
}
    duyet(head);
    //std::cout << count(head);   
    return 0;
}

int NODE(){
    Node *head = NULL;
    Node *tail = NULL;
    Node *p = NULL;
    void addlist(Node **head, int x);
    int n;
    cout << "Nhap so luong phan tu: ";
    cin >> n;
    for (int i = 0; i < n; i++)
    {
        int x;
        cout << "Nhap gia tri cho phan tu thu " << i + 1 << ": ";
        cin >> x;
        addlist(&head, x);
    }
    p = head;
    cout << "Danh sach lien ket la: ";
    while (p != NULL)
    {
        cout << p->data << " ";
        p = p->next;
    }
    // Thêm 1 node vào cuối DSLK
    void addTail(Node **head, int x);
    int x;
    cout << "Nhap gia tri can them vao cuoi DSLK: ";
    cin >> x;
    addTail(&head, x);
    p = head;
    cout << "Danh sach sau khi them vao cuoi DSLK la: ";
    while (p != NULL)
    {
        cout << p->data << " ";
        p = p->next;
    }
    cout << endl;
    // Thêm 1 node vào đầu DSLK
   void addHead(Node **head, int x); 
    cout << "Nhap gia tri can them vao dau DSLK: ";
    cin >> x;
    addHead(&head, x);
    p = head;
    cout << "Danh sach sau khi them vao dau DSLK la: ";
    printList(head);
    // Thêm 1 node vào giữa DSLK
    void addMid(Node **head, int x, int pos);
    int pos;
    cout << "Nhap vi tri can them vao DSLK: ";
    cin >> pos;
    cout << "Nhap gia tri can them vao DSLK: ";
    cin >> x;
    addMid(&head, x, pos);
    p = head;
    cout << "Danh sach sau khi them vao giua DSLK la: ";
    while (p != NULL)
    {
        cout << p->data << " ";
        p = p->next;
    }
    cout << endl;

    return 0;
}
// Xóa 1 node bất kì trong DSLK
void deleteNode(Node **head, int pos)
{
    if (*head == NULL)
    {
        return;
    }
    Node *temp = *head;
    if (pos == 0)
    {
        *head = temp->next;
        delete temp;
        return;
    }
    for (int i = 0; temp != NULL && i < pos - 1; i++)
    {
        temp = temp->next;
    }
    if (temp == NULL || temp->next == NULL)
    {
        return;
    }
    Node *next = temp->next->next;
    delete temp->next;
    temp->next = next;
}
// Menu

File: Bài thi/DSLK.py
class Node:
    def __init__(self, data=None, next=None):
        self.data = data
        self.next = next


class LinkedList:
    def __init__(self):
        self.head = None

    def traverse(self):
        current = self.head
        while current:
            print(current.data, end=" ")
            current = current.next

    def add_to_list(self, x):
        new_node = Node(x)
        if not self.head:
            self.head = new_node
        else:
            temp = self.head
            while temp.next:
                temp = temp.next
            temp.next = new_node

    def add_to_tail(self, x):
        self.add_to_list(x)

    def add_to_head(self, x):
        new_node = Node(x, self.head)
        self.head = new_node

    def add_to_mid(self, x, pos):
        new_node = Node(x)
        if pos == 0:
            new_node.next = self.head
            self.head = new_node
        else:
            temp = self.head
            for _ in range(pos - 1):
                temp = temp.next
            new_node.next = temp.next
            temp.next = new_node

    def delete_node(self, pos):
        if not self.head:
            return
        temp = self.head
        if pos == 0:
            self.head = temp.next
            return
        for _ in range(pos - 1):
            temp = temp.next
        if temp is None or temp.next is None:
            return
        temp.next = temp.next.next

    def menu(self):
        while True:
            print("\n\nCHUONG TRINH QUAN LY DANH SACH LIEN KET DON")
            print("1. Nhap du lieu")
            print("2. Them 1 node vao cuoi DSLK")
            print("3. Them 1 node vao dau DSLK")
            print("4. Them 1 node vao giua DSLK")
            print("5. Xoa 1 node trong DSLK")
            print("6. In ra DSLK")
            print("0. Ket thuc")

            luachon = int(input("Nhap vao lua chon: "))

            if luachon < 0 or luachon > 6:
                print("Lua chon khong hop le")
            elif luachon == 1:
                n = int(input("Nhap so luong phan tu: "))
                for i in range(n):
                    x = int(input(f"Nhap gia tri cho phan tu thu {i + 1}: "))
                    self.add_to_list(x)
                p = self.head
                print("Danh sach lien ket la: ", end="")
                while p:
                    print(p.data, end=" ")
                    p = p.next
            elif luachon == 2:
                x = int(input("Nhap gia tri can them vao cuoi DSLK: "))
                self.add_to_tail(x)
                p = self.head
                print("Danh sach lien ket la: ", end="")
                while p:
                    print(p.data, end=" ")
                    p = p.next
            elif luachon == 3:
                x = int(input("Nhap gia tri can them vao dau DSLK: "))
                self.add_to_head(x)
                p = self.head
                print("Danh sach lien ket la: ", end="")
                while p:
                    print(p.data, end=" ")
                    p = p.next
            elif luachon == 4:
                pos = int(input("Nhap vi tri can them vao DSLK: "))
                x = int(input("Nhap gia tri can them vao DSLK: "))
                self.add_to_mid(x, pos)
                p = self.head
                print("Danh sach lien ket la: ", end="")
                while p:
                    print(p.data, end=" ")
                    p = p.next
            elif luachon == 5:
                pos = int(input("Nhap vi tri can xoa trong DSLK: "))
                self.delete_node(pos)
                p = self.head
                print("Danh sach lien ket la: ", end="")
                while p:
                    print(p.data, end=" ")
                    p = p.next
            elif luachon == 6:
                print("Danh sach lien ket la: ", end="")
                self.traverse()
            else:
                break


def main():
    linked_list = LinkedList()
    linked_list.menu()


if __name__ == "__main__":
    main()
